### DISCUSSION TITLE
Style-Guiding Identifiers

### DISCUSSION THREAD

**Discussion by davschneller** [SeisSol/SeisSol#929] [General]
Hi everyone,

since we've now `clang-tidy` available for the whole code, there's still an issue that we could talk about:

what kind of style do we want to enforce for identifiers (e.g. class names, attributes, enum values etc.)?

For example, (private) class member variables currently either have the prefix `m_` (cf. e.g. LtsLayout), no prefix, or the suffix `_` as seen in #876 .
The question is... On what do we want to settle here? Nothing seems to be super ideal of the three. Some thoughts:

* `m_` is makes the code pretty hard to read, in my opinion.
* `_` as suffix is still bearable, even though it also decreases the readability by a bit.
* no prefix/suffix would be best, but how to deal with getters/setters then? Having the getters/setters be prefixed with `get`/`set` always also looks a bit convoluted IMO.

Another questions is e.g.: if we have a `constexpr` or `static constexpr`: should we take `camelCase` or `PascalCase`? (I'd personally prefer the latter)

Since we're at it: any opinion on header file endings or file name casing?

**Comment by sebwolf-de** [comment]
* `m_` (and also `i_`, `o_`, `l_` is horrible to read.
* suffix `_` is okay.
* no pre/suffix is preferred. A proper IDE will show you the access modifier anyways.
* getters/setters: IMHO we should only use getters/setters, if the setter has some sanity check, otherwise we can make the variable public. I don't think the `get`/`set` prefix makes things too convoluted, since they only have three letters.
* any `constexpr` should be `PascalCase`.
* I'd prefer `.cpp/.h`

**Reply by krenzland** [reply]
Completely agree with @sebwolf-de, with the exception of constexpr. IMO it should be camelCase because then we can easily switch variables/functions from const to constexpr. But that's not really a strong opinion.

**Reply by davschneller** [reply]
We could maybe differentiate between a global constexpr and a method-local constexpr—if that solves it. ... Or maybe it falls together with constants in general, let's see what `clang-tidy` gives us as options...

**Comment by davschneller** [comment]
In general, what we're mostly talking about can also be found in

https://github.com/SeisSol/SeisSol/blob/master/.clang-tidy

.

Two more things to discuss:
* what about global constants? Some do upper SNAKE_CASE here. We could also use PascalCase and reserve upper SNAKE_CASE for preprocessor defines.
* What about template parameters? Typedefs/usings? Unions?
  * Personal suggestions would be: PascalCase, PascalCase, PascalCase (no, this is not meant for emphasis, just an individual answer to each question :D). For some reason. Or alternatively a `_t` suffix? Or `T`?

**Reply by krenzland** [reply]
> what about global constants? Some do upper SNAKE_CASE here. We could also use PascalCase and reserve upper SNAKE_CASE for preprocessor defines.

Yup. Would also apply to enum (they are effectively global constants) and also due to consistency to enum class (?)

> What about template parameters? Typedefs/usings? Unions?
>     Personal suggestions would be: PascalCase, PascalCase, PascalCase (no, this is not meant for emphasis, just an individual answer to each question :D). For some reason. Or alternatively a _t suffix? Or T?

Yup. Typedef is a bit more complicated IMO. If for example I typedef int128 I wouldn't want to write Int128 instead.
I also like the _t suffix but it's probably too inconsistent.

**Reply by sebwolf-de** [reply]
I vote for PascalCase for enum and enum class.
```
switch (materialType) {
  case ANISOTROPIC: ... 
  case ELASTIC: ...
  case POROELASTIC: ...
}
```
Can you distinguish between typedefs, e.g. `using real = double` and `using Matrix = Eigen::Matrix<...>`

**Comment by davschneller** [comment]
Adding this here, we could also think of adopting e.g. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines gradually? Other than that, I'll try to come up with a suitable `clang-tidy` file hopefully pretty soon-ish.

**Comment by davschneller** [comment]
(closing; `clang-tidy` has been applied to most parts of the code by now—though some smaller issues have been ignored for now (e.g. removing all remaining `m_` prefixes, converting all methods of form `result fn(args...)` to the new `auto fn(args...) -> result` style, etc.), and maybe some applications need to be more thorough. Some included CppCoreGuideline checks are enabled as well. The remaining pieces will be addressed by #1205, #1122 and an upcoming PR)
